src/marqo/s2_inference/s2_inference.py

```diff
@@ -79,19 +83,105 @@ def _create_model_cache_key(model_name: str, device: str, model_properties: dict
     if model_cache_key not in available_models:
+        device_memory_manage(model_name, validated_model_properties, device)
         try:
-            available_models[model_cache_key] = _load_model(model_name,
-                                                            validated_model_properties, device=device)
-            logger.info(f'loaded {model_name} on device {device} with normalization={normalize_embeddings}')
+            most_recently_used_time = datetime.datetime.now()
+            available_models[model_cache_key] = {"model":_load_model(model_name,
+                                                            validated_model_properties, device=device),
+                                                 "most_recently_used_time": most_recently_used_time,
+                                                 "model_size": model_size}
+            logger.info(f'loaded {model_name} on device {device} with normalization={normalize_embeddings} at time={most_recently_used_time}.')
         except:
             raise ModelLoadError(
                 f"Unable to load model={model_name} on device={device} with normalization={normalize_embeddings}. "
                 f"If you are trying to load a custom model, "
                 f"please check that model_properties={validated_model_properties} is correct "
                 f"and the model has valid access permission. ")
+    else:
+        most_recently_used_time = datetime.datetime.now()
+        logger.debug(f'renew {model_name} on device {device} with new time={most_recently_used_time}.')
+        available_models[model_cache_key]["most_recently_used_time"] = most_recently_used_time
+
+
+def get_model_size(model_name:str, model_properties:dict) -> (int, float):
+    '''
+    Return the model size for given model
+    Note that the priorities are size_in_properties -> model_name -> model_type
+    '''
+    if "model_size" in model_properties:
+        return model_properties["model_size"]
+
+    name_info = (model_name + model_properties.get("name", "")).lower().replace("/", "-")
+    for name, size in constants.MODEL_NAME_SIZE_MAPPING.items():
+        if name in name_info:
+            return size
+
+    type = model_properties["type"]
+    return constants.MODEL_TYPE_SIZE_MAPPING.get(type, constants.DEFAULT_MODEL_SIZE)
+
+
+def device_memory_manage(model_name: str, model_properties: dict, device: str) -> bool:
+    '''
+    A function to manage the memory usage in devices when we want to load a new model
+    Args:
+        model_name: The name of the model to load
+        model_properties: The model properties of the model
+        device: The target device to laod the model
+    Returns:
+        True we have enough space for the model
+        Raise an error and return False if we can't find enough space for the model.
+    '''
+    model_size = get_model_size(model_name, model_properties)
+    if check_device_memory_status(device, model_size):
```

Comment:
This isn't threadsafe - remember this can be potentially be updated by multiple threads...